File: lib/javascript/gui.js
// Class for graphics
class GUI {
  constructor() {
    this.addButtonListener();
  }

  // add event listeners for all the buttons
  addButtonListener() {
    // add event listeners to each hint button
    this.hintButtons = document.getElementsByClassName("hints");
    for (let i = 0; i < this.hintButtons.length; i++) {
      this.hintButtons[i].addEventListener("click", () => {
        this.generate_hint(i + 1);
      });
    }
    // add event listeners to add 3 cards button
    this.add3cardsButton = document.getElementById("add-3-cards");
    this.add3cardsButton.addEventListener("click", () =>
      this.add_three_cards()
    );
    // add event listener to change players button
    this.changePlayerButton = document.getElementById("change-player");
    this.changePlayerButton.addEventListener("click", () =>
      this.change_player()
    );
    // add event listener to restart game button
    this.restartGameButton = document.getElementById("restart-game");
    this.restartGameButton.addEventListener("click", () => this.restart_game());
  }

  // switch between player1 and player2
  change_player() {
    let player_num = game.current_player.playerNum;
    switch (player_num + 1) {
      case 1:
        game.next_player = player1;
        game.current_player = player2;
        break;
      case 2:
        game.next_player = player2;
        game.current_player = player1;
        break;
    }
    // reflect who is selected on html
    game.next_player.deselect();
    game.current_player.select();
  }

  // generate hint if button is clicked
  generate_hint(hint_num) {
    // vairable to see if a set is found
    let found_set = false;
    let i = 0;
    let j = 0;
    let k = 0;
    // Iterate through each three pairs of cards
    while (i < card_deck.current_deck.length && !found_set) {
      let card1 = card_deck.current_deck[i];
      j = i + 1;
      while (j < card_deck.current_deck.length && !found_set) {
        let card2 = card_deck.current_deck[j];
        k = j + 1;
        while (k < card_deck.current_deck.length && !found_set) {
          let card3 = card_deck.current_deck[k];
          if (card_deck.is_set([card1, card2, card3])) {
            found_set = true;
          }
          k++;
        }
        j++;
      }
      i++;
    }

    if (found_set) {
      // Display the card numbers that are a set to help users in each hint
      switch (hint_num) {
        case 1:
          window.alert(`Hint 1: Consider card ${i}`);
          break;
        case 2:
          window.alert(`Hint 2: Consider card ${i} and ${j}!`);
          break;
        case 3:
          window.alert(`Hint 3: Consider card ${i}, ${j}, and ${k}!`);
          break;
        default:
          window.alert(`Error!!!`);
      }
    } else {
      window.alert(`No sets found!`);
    }
  }

  // Function to add three cards if no set is remaining
  add_three_cards() {
    //  Make sure there is cards availible to add
    if (card_deck.deck_of_cards.length != 0) {
      // Get three cards and update current deck
      for (let i = 0; i < 3; i++) {
        let card = card_deck.deck_of_cards.pop();
        card_deck.current_deck.push(card);
      }
      //  warn user that there are no more sets
      if (
        !card_deck.set_remaining(card_deck.current_deck) &&
        card_deck.deck_of_cards.length == 0
      ) {
        window.alert(`Adding 3 cards as no sets can be found`);
      }
      // Display the updated deck to GUI
      Cards.print_cards(card_deck);
    }
  }

  restart_game() {
    // hide button
    this.restartGameButton.classList.add("hide");
    window.alert("Welcome to Game of Set");

    player1 = new Player(0);
    player2 = new Player(1);
    card_deck = new Deck();
    game = new Game(player1, player2);
  }
}

File: Project2/lib/gui.rb
# Class for graphics
class GUI
  # button class for GUI
  class Button
    attr_reader :width, :height, :x, :y

    # creates a button with text using ruby2d
    def initialize(width, height, x, y, text)
      @width = x + width
      @height = y + height
      @x = x
      @y = y
      Rectangle.new x: @x, y: @y, width: width, height: height, color: 'teal'
      Text.new text, x: @x + 10, y: @y + 10, size: 20, color: 'black', z: 1
    end

    # checks if button is clicked
    def clicked?(event)
      (event.x > @x) && (event.x < @width) && (event.y > @y) && (event.y < @height)
    end

    # generate hint for player
    def generate_hint(card_deck, hint_num)
      # vairable to see if a set is found
      found_set = false
      i = 0
      # Iterate through each three pairs of cards
      while i < card_deck.current_deck.length && !found_set
        card1 = card_deck.current_deck[i]
        j = i + 1
        while j < card_deck.current_deck.length && !found_set
          card2 = card_deck.current_deck[j]
          k = j + 1
          while k < card_deck.current_deck.length && !found_set
            card3 = card_deck.current_deck[k]
            found_set = card_deck.is_set?([card1, card2, card3])
            k += 1
          end
          j += 1
        end
        i += 1
      end
      if found_set
        # Display the card numbers that are a set to help users in each hint
        puts "Hint 1: Consider card #{i}!" if hint_num == 1
        puts "Hint 2: Consider card #{i} and #{j}!" if hint_num == 2
        puts "Hint 3: Consider card #{i}, #{j}, and #{k}!" if hint_num == 3
      else
        puts 'No sets found!'
      end
    end

    # Function to add three cards if no set is remaining
    def add_three_cards()
      # Make sure there is cards availible to add
      if card_deck.deck_of_cards.length != 0
        3.times do
          # Get three cards and update current deck
          card = card_deck.deck_of_cards.pop
          card_deck.current_deck << card
        end
        # warn user that there are no more sets
        unless (card_deck.set_remaining? card_deck.current_deck) && !card_deck.deck_of_cards.length.zero?
          puts 'Adding 3 cards as no sets can be found'
        end
        # Display the updated deck to GUI
        Cards.print_cards(card_deck)
      end
    end
  end

  # actions to start the graphic interface
  def initialize
    create_buttons
  end

  attr_reader :hint1_button, :hint2_button, :hint3_button, :add_cards_button

  # creates all the buttons used for graphics
  def create_buttons
    # hint buttons
    @hint1_button = Button.new(95, 45, 750, 50, 'Hint 1')
    @hint2_button = Button.new(95, 45, 750, 100, 'Hint 2')
    @hint3_button = Button.new(95, 45, 750, 150, 'Hint 3')
    # add 3 cards button
    @add_cards_button = Button.new(130, 45, 732, 200, 'Add 3 Cards')
  end

  # replaces the card sprite at index
  def self.replace_sprite(card, card_deck, index)
    # remove sprite from gui
    delete_sprite(card_deck, index)
    # replace sprite
    card.create_card_sprite(card_deck, index)
  end

  # Delete the card sprite
  def self.delete_sprite(card_deck, index)
    card_deck.displayed_sprites[index].remove
  end
end
